Config.SkipMetadataApiCheck No Longer Working When Only EC2 Metadata Credentials

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Environment and Versions

* Terraform Executor: EC2 Instance
  * If outside Terraform Cloud/Enterprise, which operating system and version: All
* Terraform CLI version: Unreleased
* Terraform AWS Provider version: Unreleased
* Terraform Backend/Provider Configuration:

```hcl
terraform {
  backend "s3" {
    # ... other configuration ...
    skip_metadata_api_check = true
  }
}

provider "aws" {
  # ... other configuration ...
  skip_metadata_api_check = true
}
```

* AWS environment variables (if any): None

* AWS configuration files (if any): None

### Expected Behavior

EC2 Metadata API never called and those credentials never returned when `Config.SkipMetadataApiCheck` is enabled. Should return no valid credentials error.

### Actual Behavior

EC2 Metadata API is called and those credentials are returned.

### Steps to Reproduce

```go
// Additional test case in TestGetSession()
		{
			Config: &Config{
				Region:               "us-east-1",
				SkipMetadataApiCheck: true,
			},
			Description:             "skip EC2 metadata API check",
			EnableEc2MetadataServer: true,
			ExpectedError: func(err error) bool {
				// TODO: https://github.com/hashicorp/aws-sdk-go-base/pull/42
				return err.Error() == errMsgNoValidCredentialSources
			},
			ExpectedRegion: "us-east-1",
		},
```

### Debug Output

```
2020/06/03 10:37:16 [INFO] Attempting to use session-derived credentials
2020/06/03 10:37:16 [INFO] Setting custom EC2 metadata endpoint: http://127.0.0.1:65336/latest
2020/06/03 10:37:16 [DEBUG] Mock EC2 metadata server received request: /latest/meta-data/iam/security-credentials/
2020/06/03 10:37:16 [DEBUG] Mock EC2 metadata server received request: /latest/meta-data/iam/security-credentials/test_role
2020/06/03 10:37:16 [INFO] Successfully derived credentials from session
2020/06/03 10:37:16 [INFO] AWS Auth provider used: "EC2RoleProvider"
2020/06/03 10:37:16 [INFO] Setting custom STS endpoint: http://127.0.0.1:65337
2020/06/03 10:37:16 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2020/06/03 10:37:16 [DEBUG] Mock EC2 metadata server received request: /latest/meta-data/iam/security-credentials/
2020/06/03 10:37:16 [DEBUG] Mock EC2 metadata server received request: /latest/meta-data/iam/security-credentials/test_role
2020/06/03 10:37:16 [DEBUG] Received STS API "POST" request to "/": Action=GetCallerIdentity&Version=2011-06-15
--- FAIL: TestGetSession (0.00s)
    --- FAIL: TestGetSession/skip_EC2_metadata_API_check (0.00s)
        session_test.go:1030: unexpected GetSession() error: error validating provider credentials: error calling sts:GetCallerIdentity: :
            	status code: 400, request id:
```

### Fix Proposal

- Set `AWS_EC2_METADATA_DISABLED=true` environment variable if `Config.SkipMetadataApiCheck` at beginning of `GetSession()`

```go
	if c.SkipMetadataApiCheck {
		os.Setenv("AWS_EC2_METADATA_DISABLED", "true")
	}
```

- Remove `GetCredentialsFromMetadata()` entirely -- its now superseded by the default AWS Go SDK credential handling. Its endpoint customization was already migrated to `endpoints.Resolver` and the `AWS_METADATA_TIMEOUT` customization is extraneous. 

### References

- Caused by: #20

--- a/awsauth.go
+++ b/awsauth.go
@@ -13,7 +13,6 @@ import (
 	awsCredentials "github.com/aws/aws-sdk-go/aws/credentials"
 	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
 	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
-	"github.com/aws/aws-sdk-go/aws/defaults"
 	"github.com/aws/aws-sdk-go/aws/ec2metadata"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/iam"
@@ -201,84 +200,6 @@ func GetCredentialsFromSession(c *Config) (*awsCredentials.Credentials, error) {
 	return creds, nil
 }
 
-// GetCredentialsFromMetadata returns credentials derived from and ECS or ECS
-// metadata endpoint.
-func GetCredentialsFromMetadata(c *Config) (*awsCredentials.Credentials, error) {
-	log.Printf("[INFO] Attempting to use metadata-derived credentials")
-	providers := []awsCredentials.Provider{}
-
-	// Build isolated HTTP client to avoid issues with globally-shared settings
-	client := cleanhttp.DefaultClient()
-	client.Timeout = DefaultMetadataClientTimeout
-
-	const userTimeoutEnvVar = "AWS_METADATA_TIMEOUT"
-	userTimeout := os.Getenv(userTimeoutEnvVar)
-	if userTimeout != "" {
-		newTimeout, err := time.ParseDuration(userTimeout)
-		if err == nil {
-			if newTimeout.Nanoseconds() > 0 {
-				client.Timeout = newTimeout
-			} else {
-				log.Printf("[WARN] Non-positive value of %s (%s) is meaningless, ignoring", userTimeoutEnvVar, newTimeout.String())
-			}
-		} else {
-			log.Printf("[WARN] Error converting %s to time.Duration: %s", userTimeoutEnvVar, err)
-		}
-	}
-
-	log.Printf("[INFO] Setting AWS metadata API timeout to %s", client.Timeout.String())
-	cfg := &aws.Config{
-		HTTPClient: client,
-	}
-	usedEndpoint := setOptionalEndpoint(cfg)
-
-	// Add the default AWS provider for ECS Task Roles if the relevant env variable is set
-	if uri := os.Getenv("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI"); len(uri) > 0 {
-		providers = append(providers, defaults.RemoteCredProvider(*cfg, defaults.Handlers()))
-		log.Print("[INFO] ECS container credentials detected, RemoteCredProvider added to auth chain")
-	}
-
-	if !c.SkipMetadataApiCheck {
-		// Real AWS should reply to a simple metadata request.
-		// We check it actually does to ensure something else didn't just
-		// happen to be listening on the same IP:Port
-		ec2Session, err := session.NewSession(cfg)
-
-		if err != nil {
-			return nil, fmt.Errorf("error creating EC2 Metadata session: %w", err)
-		}
-
-		metadataClient := ec2metadata.New(ec2Session)
-		if metadataClient.Available() {
-			providers = append(providers, &ec2rolecreds.EC2RoleProvider{
-				Client: metadataClient,
-			})
-			log.Print("[INFO] AWS EC2 instance detected via default metadata" +
-				" API endpoint, EC2RoleProvider added to the auth chain")
-		} else {
-			if usedEndpoint == "" {
-				usedEndpoint = "default location"
-			}
-			log.Printf("[INFO] Ignoring AWS metadata API endpoint at %s "+
-				"as it doesn't return any instance-id", usedEndpoint)
-		}
-	}
-
-	// Validate the credentials before returning them
-	creds := awsCredentials.NewChainCredentials(providers)
-	cp, err := creds.Get()
-	if err != nil {
-		if IsAWSErr(err, "NoCredentialProviders", "") {
-			return nil, c.NewNoValidCredentialSourcesError(err)
-		}
-		return nil, fmt.Errorf("Error deriving credentials from metadata: %s", err)
-	}
-
-	log.Printf("[INFO] Successfully derived credentials from metadata")
-	log.Printf("[INFO] AWS Auth provider used: %q", cp.ProviderName)
-	return creds, nil
-}
-
 // GetCredentials gets credentials from the environment, shared credentials,
 // the session (which may include a credential process), or ECS/EC2 metadata endpoints.
 // GetCredentials also validates the credentials and the ability to assume a role
@@ -311,10 +232,7 @@ func GetCredentials(c *Config) (*awsCredentials.Credentials, error) {
 		if IsAWSErr(err, "NoCredentialProviders", "") {
 			creds, err = GetCredentialsFromSession(c)
 			if err != nil {
-				creds, err = GetCredentialsFromMetadata(c)
-				if err != nil {
-					return nil, err
-				}
+				return nil, err
 			}
 		} else {
 			return nil, fmt.Errorf("Error loading credentials for AWS Provider: %w", err)

--- a/session.go
+++ b/session.go
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"log"
 	"net/http"
+	"os"
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/endpoints"
@@ -65,6 +66,10 @@ func GetSessionOptions(c *Config) (*session.Options, error) {
 
 // GetSession attempts to return valid AWS Go SDK session.
 func GetSession(c *Config) (*session.Session, error) {
+	if c.SkipMetadataApiCheck {
+		os.Setenv("AWS_EC2_METADATA_DISABLED", "true")
+	}
+
 	options, err := GetSessionOptions(c)
 
 	if err != nil {

--- a/session_test.go
+++ b/session_test.go
@@ -922,7 +922,7 @@ aws_secret_access_key = DefaultSharedCredentialsSecretKey
 			},
 			Description: "session creation error",
 			ExpectedError: func(err error) bool {
-				return IsAWSErr(err, "NoCredentialProviders", "")
+				return IsAWSErr(err, "CredentialRequiresARNError", "")
 			},
 			SharedConfigurationFile: `
 [profile SharedConfigurationProfile]
@@ -941,6 +941,18 @@ source_profile = SourceSharedCredentials
 				AccessKeyID:     "StaticAccessKey",
 				ProviderName:    credentials.StaticProviderName,
 				SecretAccessKey: "StaticSecretKey",
+			},
+			ExpectedRegion: "us-east-1",
+		},
+		{
+			Config: &Config{
+				Region:               "us-east-1",
+				SkipMetadataApiCheck: true,
+			},
+			Description:             "skip EC2 metadata API check",
+			EnableEc2MetadataServer: true,
+			ExpectedError: func(err error) bool {
+				return IsNoValidCredentialSourcesError(err)
 			},
 			ExpectedRegion: "us-east-1",
 		},

